### PR TITLE
Added smooth lights

### PR DESCRIPTION
**Pull request by piotr-j**
Smooth lights are a drop in replacement for current lights. They cast rays more intelligently allowing to use less rays then comparable default light. See https://www.youtube.com/watch?v=24XHJPjpxy4 for example.

Code review would be great!

### CODE REVIEW

**Comment by rinold** [comment]
Video looks really great. But SO big... will need time.

One point - "smooth" doesn't give me much understanding :) Could you please provide some complete (brief, please, where possible) description of what and how it's different (without too much technical details - think will find that details in review). Just to understand the idea under it.

**Comment by piotr-j** [comment]
Smooth refers to how the light doesnt "jump" when objects move within its bounds. Eliminating this effect was the main goal of this project. Besides lights looking better, potential for better performance is also there in addition to some features not present in the current lights. 

One of them is lights "sleeping", not updating when only static objects are present within the bounds, resulting in extra performance. 

Jumping effect is easier to see when the video is played at .25 speed.

File: src/box2dLight/SmoothPositionalLight.java [hunk 1]
```diff
@@ -0,0 +1,714 @@
...
+	protected final Vector2 start = new Vector2();
+	protected final Rectangle aabb = new Rectangle();
+
+	protected Body body;
+	protected float bodyOffsetX;
+	protected float bodyOffsetY;
+	protected float bodyAngleOffset;
+
+	// TODO do we need these for anything?
+	protected float sin[];
```
Comment (intrigus) [inline src/box2dLight/SmoothPositionalLight.java]: Has this TODO been resolved?

File: src/box2dLight/SmoothPositionalLight.java [hunk 2]
```diff
@@ -0,0 +1,714 @@
...
+			if ((globalFilterA != null) && !globalContactFilter(fixture)) return -1;
+			if ((filterA != null) && !contactFilter(fixture)) return -1;
+			if (ignoreBody && fixture.getBody() == getBody()) return -1;
+
+			current.set(point.x, point.y, fraction);
+			return fraction;
+		}
+	};
+
+	protected float cornerOffset = 0.025f;
```
Comment (intrigus) [inline src/box2dLight/SmoothPositionalLight.java]: Specific reason for this value or just a small number?

File: src/box2dLight/SmoothPositionalLight.java [hunk 3]
```diff
@@ -0,0 +1,714 @@
...
+				}
+				break;
+			case Polygon:
+			case Chain:
+				int vc = getVertexCount(shape);
+				getVertex(fixture, vc - 1, vert1);
+				// add first corner if needed
+				if (vert1.dst2(start) <= dst2) {
+					if (currentRayNum + 2 >= rayNum) break;
+					// we dont shoot directly at the cornet, as sometimes the ray goes straight through it
```
Comment (intrigus) [inline src/box2dLight/SmoothPositionalLight.java]: Little typo "corner" instead of "cornet"

File: src/box2dLight/SmoothPositionalLight.java [hunk 4]
```diff
@@ -0,0 +1,714 @@
...
+					addRay(vert1, dst2, cornerOffset);
+				}
+				for (int i = 0; i < vc; i++) {
+					getVertex(fixture, i, vert2);
+					int found = CircleUtils.findIntersections(start, distance, vert1, vert2, tmp3, tmp4);
+					// NOTE a bit of margin, cus floats
+					// NOTE checking for dest saves as a bunch of rays for points that are not on the line
```
Comment (intrigus) [inline src/box2dLight/SmoothPositionalLight.java]: "us" instead of "as"

File: src/box2dLight/SmoothPositionalLight.java [hunk 5]
```diff
@@ -0,0 +1,714 @@
...
+						}
+						if (tmp4.dst2(start) <= dst2 + 0.001f
+								&& MathUtils.isZero(Intersector.nearestSegmentPoint(vert1, vert2, tmp4, tmp5).dst2(tmp4), 0.01f)) {
+							addRay(tmp4, dst2);
+						}
+					}
+					// also add corner
+					if (vert2.dst2(start) <= dst2) {
+						if (currentRayNum + 2 >= rayNum) break;
+						// we dont shoot directly at the cornet, as sometimes the ray goes straight through it
```
Comment (intrigus) [inline src/box2dLight/SmoothPositionalLight.java]: "corner" instead of "cornet"

File: src/box2dLight/SmoothPositionalLight.java [hunk 6]
```diff
@@ -0,0 +1,714 @@
...
+		aabb.set(start.x, start.y, 0, 0);
+		for (int i = 0; i < baseRayNum; i++) {
+			f[i] = 1f;
+			tmpEnd.x = endX[i] + start.x;
+			mx[i] = tmpEnd.x;
+			tmpEnd.y = endY[i] + start.y;
+			my[i] = tmpEnd.y;
+			extendBounds(aabb, tmpEnd.x, tmpEnd.y);
+		}
+		// TODO do we need this? might be useful cus floats
```
Comment (intrigus) [inline src/box2dLight/SmoothPositionalLight.java]: Resolved?

File: src/box2dLight/SmoothPositionalLight.java [hunk 7]
```diff
@@ -0,0 +1,714 @@
...
+		}
+	}
+
+	public void drawEdge(ShapeRenderer renderer) {
+		float x1, y1, x2, y2;
+		if (isSoft()) {
+			if (softEdgeColor != null) {
+				renderer.setColor(softEdgeColor);
+			} else {
+				renderer.setColor(1, 1, 0, .25f);
```
Comment (intrigus) [inline src/box2dLight/SmoothPositionalLight.java]: Maybe add comment like this: // semi-transparent Yellow

File: src/box2dLight/SmoothPositionalLight.java [hunk 8]
```diff
@@ -0,0 +1,714 @@
...
+				x2 = segments[(i + 8 + 4) % count];
+				y2 = segments[(i + 8 + 5) % count];
+				renderer.line(x1, y1, x2, y2);
+			}
+			lightMesh.getVertices(segments);
+		}
+		if (hardEdgeColor != null)
+			renderer.setColor(hardEdgeColor);
+		else {
+			renderer.setColor(1, 0, 0, .25f);
```
Comment (intrigus) [inline src/box2dLight/SmoothPositionalLight.java]: Same here, add comment like this: // semi-transparent Red

File: test/tests/TestApp.java [hunk 9]
```diff
@@ -0,0 +1,722 @@
...
+				if (light != null) light.remove(true);
+				switch (type) {
+				case POINT:
+					light = new DebugPointLight(rayHandler, rayCount, Color.RED, 6, 0, 0);
+					break;
+				case CONE:
+					light = new DebugConeLight(rayHandler, rayCount, Color.RED, 6, 0, 0, 0, 45);
+					break;
+				case LINE:
+					// note: default ChainLight is fixed, doesnt more or rotate
```
Comment (intrigus) [inline test/tests/TestApp.java]: "move" instead of "more"

File: test/tests/TestApp.java [hunk 10]
```diff
@@ -0,0 +1,722 @@
...
+					light = new ChainLight(rayHandler, rayCount, Color.RED, 6, 1, new float[] {
+						-1.5f, 0, 1.5f, 0});
+					break;
+				case CHAIN:
+					// note: default ChainLight is fixed, doesnt more or rotate
```
Comment (intrigus) [inline test/tests/TestApp.java]: "move" instead of "more"

File: src/box2dLight/CircleUtils.java [hunk 11]
```diff
@@ -0,0 +1,112 @@
...
+		dx = end.x - start.x;
+		dy = end.y - start.y;
+
+		A = dx * dx + dy * dy;
+		B = 2 * (dx * (start.x - center.x) + dy * (start.y - center.y));
+		C = (start.x - center.x) * (start.x - center.x) + (start.y - center.y)
+				* (start.y - center.y) - radius * radius;
+
+		det = B * B - 4 * A * C;
+		if ((A <= 0.001f) || (det < 0)) {
```
Comment (piotr-j) [inline src/box2dLight/CircleUtils.java]: Its just close enough to 0 kinda deal, epsilon if you will. The real reason is math that I cant remember at this point.

File: src/box2dLight/SmoothChainLight.java [hunk 12]
```diff
@@ -0,0 +1,275 @@
...
+import com.badlogic.gdx.math.Vector2;
+import com.badlogic.gdx.physics.box2d.Body;
+import com.badlogic.gdx.utils.Array;
+
+/**
+ * This is a composite of multiple line and cone lights to
+ * Created by PiotrJ on 16/11/2015.
+ */
+public class SmoothChainLight extends Light implements DebugLight {
+	private int side = 1;
```
Comment (piotr-j) [inline src/box2dLight/SmoothChainLight.java]: Which side the light will appear on. If you have a one chain link, a line, 1 will make the light go up, -1 will go down. Or other way around. Its same as default chain light

File: src/box2dLight/SmoothLineLight.java [hunk 13]
```diff
@@ -0,0 +1,986 @@
...
+				}
+				tmp3.set(vert1).add(-offX * (r - offsetSize), -offY * (r - offsetSize));
+				tmp4.set(vert2).add(-offX * (r - offsetSize), -offY * (r - offsetSize));
+				if (inBounds(tmp3.x, tmp3.y)) {
+					Intersector.intersectLines(tmp3, tmp4, pos, firstRay.start, tmp5);
+					Intersector.intersectLines(tmp3, tmp4, posEnd, posEndSide, tmp6);
+					addRay(tmp5, tmp6);
+				}
+				break;
+			case Polygon:
```
Comment (piotr-j) [inline src/box2dLight/SmoothLineLight.java]: Chain and polygon are handled in the same way, fallthrough

File: src/box2dLight/SmoothChainLight.java [hunk 14]
```diff
@@ -0,0 +1,275 @@
...
+
+import com.badlogic.gdx.graphics.Color;
+import com.badlogic.gdx.graphics.glutils.ShapeRenderer;
+import com.badlogic.gdx.math.MathUtils;
+import com.badlogic.gdx.math.Vector2;
+import com.badlogic.gdx.physics.box2d.Body;
+import com.badlogic.gdx.utils.Array;
+
+/**
+ * This is a composite of multiple line and cone lights to
```
Comment (piotr-j) [inline src/box2dLight/SmoothChainLight.java]: Probably, like all of them

File: src/box2dLight/SmoothConeLight.java [hunk 15]
```diff
@@ -0,0 +1,113 @@
...
+				Sort.instance().sort(rays, sorter, 0, currentRayNum);
+				// rotated back so the endpoints are in correct place
+				for (int i = 0; i < currentRayNum; i++) {
+					Ray ray = rays[i];
+					// we mark start and end, sometimes a ray at the edge sneaks past simple bounds rejection
+					if (ray == first) {
+						startRayId = i;
+					}
+					if (ray == last) {
+						endRayId = i + 1;
```
Comment (piotr-j) [inline src/box2dLight/SmoothConeLight.java]: Im sure there was a reason for that, ill check it out

File: src/box2dLight/SmoothLineLight.java [hunk 16]
```diff
@@ -0,0 +1,986 @@
...
+					fastAddRay(tmp4, tmp5);
+				}
+				vert1.set(cp).add(offX * -(r + offsetSize), offY * -(r + offsetSize));
+				if (inBounds(vert1.x, vert1.y)) {
+					Intersector.intersectLines(vert1, virtualStart, pos, firstRay.start, tmp4);
+					Intersector.intersectLines(vert1, virtualStart, posEnd, posEndSide, tmp5);
+					fastAddRay(tmp4, tmp5);
+				}
+				break;
+			case Polygon:
```
Comment (piotr-j) [inline src/box2dLight/SmoothLineLight.java]: Chain and polygon are handled in the same way, fallthrough

**Comment by piotr-j** [comment]
A lot of typos and todos and other stuff i see. Ill sort it all out tomorrow! Too bad i dont quite remember why certain values are the way they are. This code was rotting in the clouds for way too long!

**Comment by rinold** [comment]
Hmm... Dynamic rays, I've been thinking about it previously, but honestly haven't came to satisfying solution (performance). I still haven't checked all of this, but looks like you are processing the fixtures, which in case brings some performance degradation. Of course it is very-very "it depends". I feel that in quite a complex box2d environments, your solution might be much slower than used currently.

Don't think I don't  want this being merged, but we need to take care about it. I want a weighted decision - if this should or not to replace the current lights.

File: src/box2dLight/SmoothConeLight.java [hunk 17]
```diff
@@ -0,0 +1,113 @@
...
+		if (coneDegree > 179.5f) {
+			currentRayNum++;
+			// copy first one to last, so we finish the circle
+			rays[endRayId++].set(rays[startRayId]);
+		}
+	}
+
+	protected float rayAngleStep;
+	protected void setEndPoints() {
+		rayAngleStep = actualConeDeg / (baseRayNum - 1.0f);
```
Comment (rinold) [inline src/box2dLight/SmoothConeLight.java]: I'd personally prefer using the constants types related to the values we are working with. here baseRayNum is integer, and we could use just '1' here. However looking at this evaluation, with '1f' I could clearly understand that the rayAngleStep is pretending to be a float, without looking on it's declaration - makes code a bit more readable.

**Comment by piotr-j** [comment]
Its not designed to be a replacement, or even a default, but an alternative. As you said, the performance depends on a lot more variables then in the current implementation.

**Comment by uaimedna** [comment]
I would just like to say that "smooth" lights are exactly the thing that was missing from this project. The "jump" effect around the edges has prevented me from using this library multiple times in the past. The video looks great and I can't wait to get my hands on the library when this is merged. Great work!

**Pull request merged**